### PR TITLE
 Add az_cas module to install and override cas defaults specific to the University of Arizona.

### LINKED ISSUE #52: Add CAS module support
---
name: Feature request
about:  Add az_cas module to override default CAS settings unique to the University of Arizona.
labels: 'enhancement'
assignees: 'trackleft'

---
## Motivation
As an affiliate of the University of Arizona I would like to be able to log into an az_quickstart site via the centralized authentication system.

### Is your feature request related to a problem? Please describe.
No

### Describe the solution you'd like
CAS module installs and az_cas module overrides the initial settings.

### PR DESCRIPTION
**Pull request by trackleft**
## Description
This pull request renames https://github.com/az-digital/az_quickstart/pull/13
Closes #52

## Related Issue
https://github.com/az-digital/az_quickstart/issues/52

## How Has This Been Tested?
Built az_quickstart normally, checked if modules installed:
cas
externalauth
az_cas

Check if default config is overridden with az_cas settings.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (borrowing heavily from https://bitbucket.org/ua_drupal/ua_cas/src/7.x-1.x/README.md)
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Consider excluding CAS module from drupal-check temporarily
- [x] Create d.o. issue for CAS module PHPunit deprecation warnings
- [x] resolve merge conflict(s)
- [x] update composer.json white space to use 4 spaces instead of 2

Opened an issue on Drupal.org regarding drupal 9 readiness errors.
https://www.drupal.org/project/cas/issues/3110204

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -25,13 +25,18 @@
...
+        "drupal/externalauth": "1.2.0"
     },
     "extra": {
       "patches": {
           "drupal/bootstrap_barrio": {
               "missing theme config schema": "https://www.drupal.org/files/issues/2019-11-15/improve-theme-settings-remove-deprecated-code-3037780-22.patch"
           }
+      },
+      "installer-paths": {
+        "modules/contrib/{$name}": ["type:drupal-module"]
```
Comment (joeparsons) [inline composer.json:39]: I think we want to rely on the `installer-paths` config in the scaffolding repo and not add anything to this repo's composer.json.

**Comment by trackleft** [comment]
Looks like I either need to update cas or reroll the patches to fit into the version we are using.

**Comment by trackleft** [comment]
To test, ensure all settings are correct here: https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-53.probo.build/user/login?destination=admin/config/people/cas
username: azadmin
password: probo

File: composer.json [hunk 2]
```diff
@@ -25,15 +25,23 @@
         "cweagans/composer-patches": "1.6.7",
         "drupal/core-recommended": "8.8.4",
         "drush/drush": "9.7.1",
-        "drupal/background_image_formatter": "1.4",
         "drupal/bootstrap_barrio": "4.22",
+        "drupal/cas": "1.6.0",
+        "drupal/externalauth": "1.2.0",
+        "drupal/background_image_formatter": "1.4",
```
Comment (joeparsons) [inline composer.json:31]: Nitpick: I'd be preferable to keep these in alphabetical order. We could add this to our composer.json (like we have in our scaffolding repo) to make these easier to do automatically:

"config": {
    "sort-packages": true
},

File: modules/custom/az_cas/config/install/cas.settings.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  message_no_local_account: 'You do not have an account on this website. Please contact a site administrator.'
+  message_subscriber_denied_reg: 'You do not have access to log in to this website. Please contact a site administrator if you believe you should have access.'
+  message_account_blocked: 'Your account is blocked or has not been activated. Please contact a site administrator.'
+  message_subscriber_denied_login: 'You do not have access to log in to this website. Please contact a site administrator if you believe you should have access.'
+  message_username_already_exists: 'An account on this website with your username already exists. Please contact a site administrator.'
+advanced:
+  debug_log: false
+  connection_timeout: 10
+login_link_enabled: false
+login_link_label: 'CAS Login'
```
Comment (joeparsons) [inline modules/custom/az_cas/config/install/cas.settings.yml:51]: In D7 we changed this to be `'Log in using UA WebAuth'`.

File: modules/custom/az_cas/config/install/cas.settings.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+login_success_message: ''
```
Comment (joeparsons) [inline modules/custom/az_cas/config/install/cas.settings.yml:52]: In D7 we changed this to be `'Logged in via UA WebAuth as %cas_username.'`.

File: modules/custom/az_cas/config/install/cas.settings.yml (hunk 3 above)
Comment (trackleft) [reply modules/custom/az_cas/config/install/cas.settings.yml:51]: Log in using University of Arizona WebAuth

File: modules/custom/az_cas/config/install/cas.settings.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  single_logout_session_lifetime: 25
+proxy:
+  initialize: false
+  can_be_proxied: false
+  proxy_chains: ''
+user_accounts:
+  prevent_normal_login: true
+  auto_register: false
+  email_assignment_strategy: 0
+  email_hostname: ''
```
Comment (joeparsons) [inline modules/custom/az_cas/config/install/cas.settings.yml:34]: In D7 we used `email.arizona.edu` for this.

Comment (trackleft) [reply modules/custom/az_cas/config/install/cas.settings.yml:34]: D8 should be arizona.edu

Comment (joshuasosa) [reply modules/custom/az_cas/config/install/cas.settings.yml:34]: Is it just arizona.edu now or does it still need the email. part?

File: modules/custom/az_cas/config/install/cas.settings.yml (hunk 3 above)
Comment (joshuasosa) [reply modules/custom/az_cas/config/install/cas.settings.yml:51]: Or UArizona :)

File: modules/custom/az_cas/config/install/cas.settings.yml (hunk 5 above)
Comment (joeparsons) [reply modules/custom/az_cas/config/install/cas.settings.yml:34]: I think we can just use `arizona.edu` now.

Comment (trackleft) [reply modules/custom/az_cas/config/install/cas.settings.yml:34]: You can now use just @arizona.edu as far as I know.

**Comment by trackleft** [comment]
Thanks for reviewing, this one has been in the pipeline for 3 years ha, maybe longer.

**Comment by trackleft** [comment]
also wouldn't mind suggestions on tests to make

**Comment by joeparsons** [comment]
Follow-up task: investigate if az_core is really needed as a dependency for az_cas.

**Comment by joeparsons** [comment]
Followup ticket: Add ua_cas configuration settings for disabling `user/login`, `user/register`, and `user/reset` routes.

**Comment by joeparsons** [comment]
Followup ticket: Add default values for new CAS module `prevent normal login error message` and `restrict password management error message` settings.

**Pull request merged**